llvm/test/CodeGen/AArch64/sve2-histcnt.ll

```diff
@@ -267,5 +267,79 @@ define void @histogram_i16_8_lane(ptr %base, <vscale x 8 x i32> %indices, i16 %i
   ret void
 }
 
+define void @histogram_i32_zextend(ptr %base, <vscale x 4 x i32> %indices, <vscale x 4 x i1> %mask) #0 {
```

Comment:
It might be worth testing a few more index types maybe `i8` and `i16` to check those cases are working too :thinking:

Reply:
I didn't add any more tests because the other tests only use i32 and i64 for the index types. They only vary the type returned by the getelementptr instruction. Do you think I should:
-Try adding more tests using i32 indices but varying the pointer types 
-Try adding more index types like i8 and i16 that will be extended, then add more tests to the original set with i8 and i16 index types but have them not be extended?

Reply:
I'm concerned that maybe there's a reason for that (like `histcnt` only works for i32 or i64 indices -- I've not checked that myself). But maybe there's cases like below, where you may want to do something different if `i8` indices are not directly supported by `histcnt`. Maybe then you'd want to extend to the smallest legal type (so extend to i32 rather than i64).
```
%extended = zext <vscale x 4 x i8> %indices to <vscale x 4 x i64>
%buckets = getelementptr i32, ptr %base, <vscale x 4 x i64> %extended
call void @llvm.experimental.vector.histogram.add.nxv4p0.i32(<vscale x 4 x ptr> %buckets, i32 1, <vscale x 4 x i1> %mask)
```

Reply:
Yes this test fails with the following message:
`llc: /home/jamche01/LLVMGerrit/organic-llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:2763: llvm::SDValue llvm::DAGTypeLegalizer::PromoteIntOp_VECTOR_HISTOGRAM(llvm::SDNode*, unsigned int): Assertion OpNo == 1 && "Unexpected operand for promotion"' failed.`

Should I try and change LegalizeInteger type so that it can also promote `%buckets` (maybe by adding a `|| OpNo == 0` to the assertion)? Or would this maybe be unexpected behaviour, and should I just change my solution so the type isn't truncated to its original if the original isn't i32?

Reply:
I think [here](https://github.com/llvm/llvm-project/blob/7c61336bbcb18f0c2bbda6979df13f3a81ef270d/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L24094) you want to check what the type of `ExtOp`:

- If it's an `i32`, you can use it directly as the source of the histogram operation, no worries
- If it's < `i32` (e.g. `i8`) and the target type is `i32`, do nothing
- If it's < `i32` (e.g. `i8`) and the target type is > `i32` replace it with a smaller extend (i.e. only extend to `i32`)

Maybe double check with @huntergr-arm though in case there's something else you could do :slightly_smiling_face:

Reply:
Yes, that's pretty much what we want -- if it's possible to use a 32b index type, we'd prefer to do so because we can process twice as much data that way. My idea was to reuse the existing logic for gathers and scatters (which does the same thing for the same reasons), and I believe handles more cases than just extends.

But I'll take incremental progress with just supporting extends to begin with :)